### PR TITLE
[Core] Fix Streaming generator concurrency bug

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

The ObjectRefStream APIs are used by 2 threads. The main thread & IO threads.

While we are protecting the shared data using mutex, there are parts of object ref stream APIs that don't hold a lock (particularly here https://github.com/ray-project/ray/blob/62fc2293821ce76765b2d82c257c60e86609a06d/src/ray/core_worker/task_manager.cc#L568), which causes lots of random concurrency bugs. We point fixed a couple of them, but there are still many edge cases if any of APIs from the main thread is called between this timing. E.g., issues Hao and found was the case when PeekObjectRefStream is called in this timing, but it is also possible TryReadNextItem or DeleteObjectRefStream can cause issues.

Instead of fixing all potential concurrency problems, I instead make sure the lock is held e2e while it is running APIs from the IO thread. For example, by holding a lock from the beg to the end of `HandleReportGeneratorItemReturns`, we can avoid having concurrency issues like this.

When we did this, there was one edge case that caused deadlock. It is when I hold a gil while running APIs from a main thread (PeekObjectRefStream) + using async client. Looks like when a object is created from `HandleReportGeneratorItemReturns`, it calls the python code directly from the IO thread, which requires a GIL, but GIL was already held by a thread that called `PeekObjectRefStream`. I fixed this issue by simply not holding a gil when object ref stream APIs are called from a main thread. 
## Related issue number

https://github.com/ray-project/ray/issues/38232

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/task_manager.cc [hunk 1]
```diff
@@ -470,43 +470,41 @@ Status TaskManager::TryReadObjectRefStream(const ObjectID &generator_id,
...
-           "created "
-           "and not removed.";
-    next_object_id = stream_it->second.PeekNextItem();
-  }
+  absl::MutexLock lock(&objet_ref_stream_ops_mu_);
+  auto stream_it = object_ref_streams_.find(generator_id);
+  RAY_CHECK(stream_it != object_ref_streams_.end())
+      << "PeekObjectRefStream API can be used only when the stream has been "
+         "created "
+         "and not removed.";
```
Comment (raulchen) [inline src/ray/core_worker/task_manager.cc:478]:
```suggestion
         "created and not removed.";
```

**Review by raulchen — APPROVED** [review]
LGTM. And I've verified the fix with a data stress test.

**Comment by rkooo567** [comment]
(No perf issue + ready to merge it. I will merge it in a couple hours)

**Comment by dzikrisyairozi** [comment]
very good fix

**Pull request merged**